File: lcs.py
def longest_common_subsequence(X, Y):
    m = len(X)
    n = len(Y)

    # Create a DP table to store lengths of LCS
    dp = [[0] * (n + 1) for _ in range(m + 1)]

    # Build the dp table from bottom up
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if X[i - 1] == Y[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

    # Reconstruct the LCS from the dp table
    index = dp[m][n]
    lcs = [''] * index

    i, j = m, n
    while i > 0 and j > 0:
        if X[i - 1] == Y[j - 1]:
            lcs[index - 1] = X[i - 1]
            i -= 1
            j -= 1
            index -= 1
        elif dp[i - 1][j] > dp[i][j - 1]:
            i -= 1
        else:
            j -= 1

    return ''.join(lcs)


if __name__ == "__main__":
    X = "AGGTAB"
    Y = "GXTXAYB"
    print(f"Longest Common Subsequence of '{X}' and '{Y}' is '{longest_common_subsequence(X, Y)}'")

File: lcs.js
/*
 * Longest Common Subsequence (LCS) Dynamic Programming Implementation
 *
 * This script defines a function to compute the length of the LCS between two strings
 * and also reconstructs one LCS sequence.
 *
 * Usage example:
 *   node lcs.js
 */

function longestCommonSubsequence(text1, text2) {
  const m = text1.length;
  const n = text2.length;
  // dp[i][j] will hold the length of LCS of text1[0..i-1] and text2[0..j-1]
  const dp = Array(m + 1)
    .fill(null)
    .map(() => Array(n + 1).fill(0));

  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      if (text1[i - 1] === text2[j - 1]) {
        dp[i][j] = dp[i - 1][j - 1] + 1;
      } else {
        dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
      }
    }
  }

  // Reconstruct one LCS from the dp table
  let i = m, j = n;
  let lcs = [];
  while (i > 0 && j > 0) {
    if (text1[i - 1] === text2[j - 1]) {
      lcs.push(text1[i - 1]);
      i--;
      j--;
    } else if (dp[i - 1][j] > dp[i][j - 1]) {
      i--;
    } else {
      j--;
    }
  }

  lcs = lcs.reverse().join('');

  return { length: dp[m][n], sequence: lcs };
}

// Example usage
const str1 = 'AGGTAB';
const str2 = 'GXTXAYB';
const result = longestCommonSubsequence(str1, str2);
console.log(`Longest Common Subsequence length: ${result.length}`);
console.log(`Longest Common Subsequence: ${result.sequence}`);

module.exports = { longestCommonSubsequence };